homeassistant/components/tolo/number.py

```diff
@@ -0,0 +1,111 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ToloSaunaCoordinatorEntity, ToloSaunaUpdateCoordinator
+from .const import DOMAIN, FAN_TIMER_MAX, POWER_TIMER_MAX, SALT_BATH_TIMER_MAX
+
+
+@dataclass
+class ToloNumberEntityDescriptionBase:
+    """Required values when describing TOLO Number entities."""
+
+    getter: Callable[[SettingsInfo], int | None]
+    setter: Callable[[ToloClient, int | None], Any]
+
+
+@dataclass
+class ToloNumberEntityDescription(
+    NumberEntityDescription, ToloNumberEntityDescriptionBase
+):
+    """Class describing TOLO Number entities."""
+
+    entity_category = EntityCategory.CONFIG
+    min_value = 0
+    step = 1
+
+
+NUMBERS = (
+    ToloNumberEntityDescription(
+        key="power_timer",
+        icon="mdi:power-settings",
+        name="Power Timer",
+        unit_of_measurement=TIME_MINUTES,
+        max_value=POWER_TIMER_MAX,
+        getter=lambda settings: settings.power_timer,
+        setter=lambda client, value: client.set_power_timer(value),
+    ),
+    ToloNumberEntityDescription(
+        key="salt_bath_timer",
+        icon="mdi:shaker-outline",
+        name="Salt Bath Timer",
+        unit_of_measurement=TIME_MINUTES,
+        max_value=SALT_BATH_TIMER_MAX,
+        getter=lambda settings: settings.salt_bath_timer,
+        setter=lambda client, value: client.set_salt_bath_timer(value),
+    ),
+    ToloNumberEntityDescription(
+        key="fan_timer",
+        icon="mdi:fan-auto",
+        name="Fan Timer",
+        unit_of_measurement=TIME_MINUTES,
+        max_value=FAN_TIMER_MAX,
+        getter=lambda settings: settings.fan_timer,
+        setter=lambda client, value: client.set_fan_timer(value),
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up number controls for TOLO Sauna."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [ToloNumberEntity(coordinator, entry, description) for description in NUMBERS]
+    )
+
+
+class ToloNumberEntity(ToloSaunaCoordinatorEntity, NumberEntity):
+    """TOLO Number entity."""
+
+    entity_description: ToloNumberEntityDescription
+
+    def __init__(
+        self,
+        coordinator: ToloSaunaUpdateCoordinator,
+        entry: ConfigEntry,
+        entity_description: ToloNumberEntityDescription,
+    ) -> None:
+        """Initialize TOLO Number entity."""
+        super().__init__(coordinator, entry)
+        self.entity_description = entity_description
+        self._attr_unique_id = f"{entry.entry_id}_{entity_description.key}"
+
+    @property
+    def value(self) -> float:
+        """Return the value of this TOLO Number entity."""
+        return self.entity_description.getter(self.coordinator.data.settings) or 0
```

Comment:
Shouldn't it return `None` when the value is unknown?

```suggestion
    @property
    def value(self) -> float | None:
        """Return the value of this TOLO Number entity."""
        return self.entity_description.getter(self.coordinator.data.settings)
```

Reply:
`self.coordinator.data.settings == None` means disabled, not unknown. After a chat in Discord with some devs, they suggested to use the 0 of the number entity for disabling a timeout instead of having a complete seperate switch entity for that. Therefore: It's correct the way it currently is.

The complete situation: The actual TOLO hardware returns its timer configuration in a single unsigned byte per timer, representing the actual value of the timer setting. Each timer can also be disabled, for some timers the special value is 255, for some 61 (don't ask me why please...). The `tololib` package unifies that by using `None` for disabling the timeout and the actual integer value (with a range validation) for the value. This also implies: If a timeout is disabled, the numeric timeout value is lost - which looks strange in Home Assistant if you have e.g. a switch for enabling/disabling and a number element for the timeout value. Disabling the timeout would set the special value (e.g. 61) on the device, the previous timeout is lost. Re-enabling would then require either to use a default value (which seems strange to the user that disabling/enabling changes the timeout value) or to store the timeout value somewhere - where two elements (number and switch) can access it. I discussed that in Discord with other devs and the idea was to just just number, without switch, and treat the 0 as disable value. Then, if the user disables the timeout, he changes the number element by himself and is not changed by strange magic without touching it...

For any suggestion how I can solve that even better, I'm open for ideas (here or in Discord). Otherwise, for now, I would just leave it that way.

Reply:
> self.coordinator.data.settings == None means disabled, not unknown.

That is not what this line there is now is testing:

```py
return self.entity_description.getter(self.coordinator.data.settings) or 0
```

This tests the result of getter to be None and if so, return 0.
(Which is different from checking if `self.coordinator.data.settings` is None or not.

if `self.coordinator.data.settings` is missing, the sensor should become unavailable instead?

Default to 0 is not correct behavior, as it is an assumption. Thus it should be either unknown or unavailable.

Reply:
I'm not sure if I'm missing something here, so I just try to bring some more arguments. If I'm on the wrong path, please let me know, sorry for discussion and thanks for the patience! (For me, it's more about learning the intended way to go)

So, again: We're talking about controlling the power timer (and some more, e.g. fan, but the discussion is completely analoug) of a TOLO device using HA entities. The device itsself has also only a single "property" (byte) for controlling if the timer is disabled or has a value (e.g., `30`: timer of 30 minutes, `61`: timer disabled). Since the special value `61` has no sense to be treaten in HA, the `tololib` package cares about that, and maps a disabled timer to the special value `None` - it does not mean the timer is unavailable, it just means it's disabled and you can enable it if you want.

Now, how to control this with HA? The obvious choice for setting a number would be a `Number` entity. However, a number entity does not have a special value like `None` or a special enabled/disabled feature, therefore two choices:
  * Introduce switch for enablind/disabling timers. This raises the question: When a timer is disabled and is getting enabled, which value to set it? There is no value coming from the device (because it's sending `61` then respectively `tololib` returns `None` which indicates disabled timer). This would require a lot of messing around with RestoreEntity, to somehow store the last set timer value - I tried and the code looks really ugly as well as the user experience is.
  * (as suggested by some devs in Discord, where I discussed this issue): Remove the switch, just use the number, and since all timers go from 1 to 60 (ok, there is one exception coming, which goes to 8...), use the number entity's `0` to mark *disabled* state. That's the solution that I went for, since it improves user experience a lot, simplifies the code and is much closed to the actual device, which also does not offer two different options (neither in the hardware control panel nor the very strange Android app).

> `self.coordinator.data.settings == None`

This is not part of the code, but only of my previous comment - and completely wrong. Should have been `self.coordinator.data.settings.power_timer == None`, sorry for the confusion. The code never tests `self.coordinator.data.settings == None`. `self.coordinator.data.settings.power_timer == None` means, power_timer is disabled.

Having the number-without-switch solution (as pointed out above), the number element should never become unavailable (except of course the whole device becomes unavailable), since timer can be set (enabled with a value or disabled by putting it to a "special" value) at any time.

This requires a translation from the special value `None` marking disabled to a "special" value of the `Number` entity, for which I'm taking the `0`, as suggested by the community.

Now, there are two ways to realize that: Requiring the getter definition to only return int, that would make it to
```python
getter=lambda settings: settings.power_timer or 0
```
The `or 0` part would than be used in all `ToloNumberEntityDescription`s.

Or allow the getter to also return `None`, and doing the translation to `0` somewhere else, which removes the duplicated `or 0` logic. All timers behave here the same, they return `None` if they want to indicate, that the timer is currently disabled, so the device will run indefinitely (or fan etc...). So I decided to have the translation from `None` to `0` in a more central place. Also, the translaction from `0` back to `None` when a new number value is set is done in the `ToloNumberEntity` class, and having both translations from `tololib` semantics to HA semantics and back in the same place is good style I think.

Do you get my point(s) here? Or am I missing something?